File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/E._Компоненты_связности/solution.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <string>
#include <algorithm>


void dfs(int& source, std::unordered_map<int, std::vector<int>>& adjList, std::vector<int>& colors, int& componentColor, std::vector<std::vector<int>>& components) {
    std::vector<int> callStack = {source};
    while (!callStack.empty()) {
        int vertex = callStack[callStack.size() - 1];
        callStack.pop_back();
        if (colors[vertex] == -1){
            components[componentColor].push_back(vertex);
            colors[vertex] = componentColor;
        }
        for (int& neighbor : adjList[vertex]) {
            if (colors[neighbor] == -1) callStack.push_back(neighbor);
        }
    }
}



void solution() {
    /*
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    */
    int n, m;
    std::cin >> n >> m;
    std::unordered_map<int, std::vector<int>> adjList;
    for (int i = 0; i < m; ++i) {
        int before, after;
        std::cin >> before >> after;
        adjList[before].push_back(after);
        adjList[after].push_back(before);
    }
    std::vector<int> colors (n + 1, -1);
    std::vector<std::vector<int>> components;
    int componentColor = 0;
    for (int vertex = 1; vertex <= n; ++vertex) {
        if (colors[vertex] == -1) {
            components.push_back({});
            dfs(vertex, adjList, colors, componentColor, components);
            ++componentColor;
        }
    }
    for (std::vector<int>& component : components) {
        std::sort(component.begin(), component.end());
    }
    std::sort(components.begin(), components.end());
    std::cout << componentColor << "\n";
    for (std::vector<int>& component : components) {
        for (int idx = 0; idx < component.size(); ++idx) {
            if (idx) std::cout << " ";
            std::cout << component[idx];
        }
        std::cout << "\n";
    }
}


int main() {
    solution();
}

File: Yandex/Practicum/Algorightms_and_Data_Structures/6_sprint/E._Компоненты_связности/solution.py
import sys


def dfs(source: int, adj_list: dict[int, list[int]], colors: list[int], components_color: int, components: list[list[int]]) -> None:
    call_stack: list[int] = [source]
    while call_stack:
        vertex: int = call_stack.pop()
        if colors[vertex] == -1:
            colors[vertex] = components_color
            components[components_color].append(vertex)
            for neighbor in adj_list.get(vertex, []):
                if colors[neighbor] == -1: call_stack.append(neighbor)


def solution() -> None:
    """
    Time Complexity O(V + E)
    Memory Complexity O(V + E)
    """
    n, m = map(int, sys.stdin.readline().rstrip().split())
    adj_list: dict[int, list[int]] = dict()
    for _ in range(m):
        before, after = map(int, sys.stdin.readline().rstrip().split())
        if before not in adj_list: adj_list[before] = []
        adj_list[before].append(after)
        if after not in adj_list: adj_list[after] = []
        adj_list[after].append(before)
    colors: list[int] = [-1] * (n + 1)
    components: list[list[int]] = []
    components_color: int = 0
    for vertex in range(1, n + 1):
        if colors[vertex] == -1:
            components.append([])
            dfs(vertex, adj_list, colors, components_color, components)
            components_color += 1
    for component in components: component.sort()
    components.sort()
    sys.stdout.write(str(components_color) + '\n')
    for component in components:
        sys.stdout.write(' '.join(str(vertex) for vertex in component) + '\n')


if __name__ == '__main__':
    solution()